python/tk_3de4/menu_generation.py

```diff
@@ -0,0 +1,216 @@
+
+import sgtk
+from sgtk.platform.qt import QtGui, QtCore
+
+
+class MenuGenerator(object):
+    """
+    Menu generation functionality for 3DE4
+    """
+    root_ui_item = "Main Window::Shotgun"
+
+    def __init__(self, engine):
+        """
+        Initialise the class.
+
+        :param engine: The shotgun engine instance.
+        """
+        self._engine = engine
+        self._current_menu_index = count()
+        self.custom_scripts_dir_path = os.environ["TK_3DE4_MENU_DIR"]
+
+    ##########################################################################################
+    # public methods
+
+    @property
+    def logger(self):
+        """
+        Get the logger instance.
+
+        :returns: The logger associated to the engine.
+        """
+        return self._engine.logger
+
+    def create_menu(self):
+        """
+        Render the entire Shotgun menu.
+        In order to have commands enable/disable themselves based on the enable_callback,
+        re-create the menu items every time.
+        """
+        self.logger.info("Creating Shotgun menu...")
+
+        # Get temp folder path and create it if needed.
+        if not os.path.isdir(self.custom_scripts_dir_path):
+            self.logger.debug("Creating menu directory %s", self.custom_scripts_dir_path)
+            os.makedirs(self.custom_scripts_dir_path)
+        # Clear it.
+        for item in os.listdir(self.custom_scripts_dir_path):
+            self.logger.debug("Removing menu file: %s", item)
+            os.remove(os.path.join(self.custom_scripts_dir_path, item))
+
+        menu_items = []
+        for cmd_name, cmd_details in self._engine.commands.items():
+             menu_items.append(AppCommand(cmd_name, cmd_details))
+
+        ctx = self._engine.context
+        ctx_name = str(ctx)
+        # Adding a space to name will put it at the top of the menu as it's alphabetical
+        ctx_menu_name = "{}:: {}".format(MenuGenerator.root_ui_item, ctx_name)
+
+        other_menu_items = defaultdict(list)
+        favourites = []
+
+        for cmd in menu_items:
+            if cmd.get_type() == "context_menu":
+                self.logger.debug("Adding %s to context menu", cmd.name)
+                self._add_command_to_menu(cmd, ctx_menu_name)
+            else:
+                other_menu_items[cmd.get_app_name()].append(cmd)
+            if cmd.favourite:
+                favourites.append(cmd)
+
+        for cmd in favourites:
+            self.logger.debug("Adding %s to favourites", cmd.name)
+            self._add_command_to_menu(cmd, MenuGenerator.root_ui_item, favourite=True)
+
+        for app, cmds in other_menu_items.items():
+            if len(cmds) == 1:
+                if cmds[0].favourite:  # cmd has been added to favourites, so no need to add it again
+                    continue
+                parent_menu = MenuGenerator.root_ui_item
+            else:
+                parent_menu = "{}::{}".format(MenuGenerator.root_ui_item, app)
```

Comment:
Lot's of `::` being used, maybe you can refactor to have a class variable like `MENU_SEP = "::"`

Then you do change these to something like
```python
                parent_menu = self.MENU_SEP.join([MenuGenerator.root_ui_item, app])
```